### ISSUE TITLE
[webextension] Allow dynamic manifest.json

### ISSUE THREAD

**Issue by ngustavo** [parcel-bundler/parcel#8138]
# 🙋 feature request

Everytime I update the extension, I must manually update manifest.json. It would be nice if I could use a manifest.js and get the version and other info directly from package.json.

## 🤔 Expected Behavior

Allow usage of manifest.js or add the option to get info from package.json.

## 😯 Current Behavior

It only accepts a manifest.json

## 💁 Possible Solution

Expect besides a JSON file, a JS object, a default export of JSON string or read from package.json.

## 🔦 Context

I'm trying to automate some upgrade tasks and I don't see any way of achieving this.

## 💻 Examples

This is possible with Rollup and Webpack:
https://github.com/crxjs/chrome-extension-tools

**Comment by louisgv** [comment]
I'm working on a browser extension framework that uses Parcel underneath, and it basically does exactly what you described (i.e, the manifest is generated for you and the version and so on are taken from package.json). Feel free to check out the [docs](https://docs.plasmo.com/) :D

**Comment by mohsen1** [comment]
This could be very useful for me. At the moment I have a a little pre-build script that writes `manifest.json` to disk before building.

**Comment by fregante** [comment]
Accepting `manifest.json.js` sounds like a good idea. Having a dynamic output covers a lot of cases (development-only settings, per-browser bundles, testing, etc)

**Comment by devongovett** [comment]
This would make caching and watch mode a lot more difficult. JS files can produce non-deterministic results, or may have dependencies on other files, npm packages, environment variables, etc. This is all very hard to track correctly, which is why we generally discourage JS-based config files. I would be hesitant to add another one.

**Comment by fregante** [comment]
How is this different from any other file (as far as the dependencies are concerned)? 

```diff
// index.js
- import('./alpha.js')
+ import('./beta.js')
```


```diff
// manifest.json.js
export default {
	background: {
		scripts: [
-			"bg.js"
+			"background.js"
		]
	}
}
```

They both change the graph. The only difference is that `index.js` appears "as is" in the `dist` whereas `manifest.json` appears "as exported"

**Comment by devongovett** [comment]
Because it isn't bundled, it's executed and the _results_ are written. So anything that might change those results needs to invalidate the cache, not just the file itself.

**Comment by ngustavo** [comment]
Maybe evaluate that JS on load and turn it into a string object or JSON? That way we could still read environment variables and other configurations before bundling.

**Comment by mischnic** [comment]
There could be some kind of JSON templating transformer that replaces `{{something}}` with the actual result (so this wouldn't be specific to manifest.json, just any JSON files)

**Comment by tristan-f-r** [comment]
With the templating system, it may be a good idea to use enviornment variables (it turns out that [NPM has env variables for package versions, names, and other fields](https://github.com/parcel-bundler/parcel/issues/8470#issuecomment-1413763441)) - maybe a seperate issue / discussion should be raised for this new transformer?

**Comment by ngustavo** [comment]
Is it possible to get env vars on package.json though?

**Comment by ngustavo** [comment]
If you use env vars, you'll have to sync your package.json version manually afterall. And BTW you can use `npm version patch` to increment version. I'd say it's the best approach to get the version from the root package.json.

**Comment by 101arrowz** [comment]
Part of the templating issue can be solved by using a custom transformer. Previously there were issues because the entry asset had to be `manifest.json`, but that should be fixed in this PR: #9068
